File: leetcode-solution/动态规划/139.单词拆分.ts
/*
 * @lc app=leetcode.cn id=139 lang=typescript
 *
 * [139] 单词拆分
 */

// @lc code=start
function wordBreak(word: string, wordDict: string[]): boolean {
  const n = word.length;
  const maxlen = wordDict.reduce((max, word) => Math.max(max, word.length), 0);
  const wordSet = new Set(wordDict);
  const dp = new Array(n + 1).fill(false);
  dp[0] = true;

  for (let i = 1; i <= n; i++) {
    for (let j = Math.max(0, i - maxlen); j < i; j++) {
      if (wordSet.has(word.slice(j, i)) && dp[j]) {
        dp[i] = true;
        break;
      }
    }
  }
  return dp[n];
};
// @lc code=end

File: leetcode-solution/动态规划/139.单词拆分.py
#
# @lc app=leetcode.cn id=139 lang=python3
#
# [139] 单词拆分
#

# @lc code=start
class Solution:
    def wordBreak(self, word: str, wordDict: list) -> bool:
    # Convert the word dictionary into a set for O(1) lookups
        n = len(word)
        maxlen = max(len(w) for w in wordDict) if wordDict else 0
        # Find the maximum word length in the dictionary
        wordSet = set(wordDict)
        dp = [False] * (n + 1)
        dp[0] = True
        # dp[i] represents whether the first i characters can be segmented
        # Base case: empty string can always be segmented
        for i in range(1, n + 1):
            for j in range(max(0, i - maxlen), i):
            # Check all possible segmentation points
                # Only check up to max_len characters back, or to the beginning
                if word[j:i] in wordSet and dp[j]:
                    dp[i] = True
                    break
        return dp[n]      
# @lc code=end
